pkg/apis/network/v1/egressrouter_types.go

```diff
@@ -0,0 +1,267 @@
+	// +kubebuilder:validation:Enum="Bridge";"Private";"VEPA";"Passthru"
+	// +kubebuilder:default:="Bridge"
+	Mode MacvlanMode `json:"mode"`
+
+	// Name of the master interface. Need not be specified if it can be inferred from the IP address.
+	// +kubebuilder:validation:MinLength=1
+	// +kubebuilder:validation:Optional
+	// +optional
+	Master string `json:"master,omitempty"`
+}
+
+// ProtocolType defines the protocol types that are supported
+type ProtocolType string
+
+const (
+	// ProtocolTypeTCP refers to the TCP protocol
+	ProtocolTypeTCP ProtocolType = "TCP"
+
+	// ProtocolTypeUDP refers to the UDP protocol
+	ProtocolTypeUDP ProtocolType = "UDP"
+
+	// ProtocolTypeSCTP refers to the SCTP protocol
+	ProtocolTypeSCTP ProtocolType = "SCTP"
+)
+
+// EgressRouterModeConfig represents a single mode's configuration.
+// Mode must be specified along with exactly one "Config" that matches the mode.
+// Each config consists of parameters used to define the remote destination that the router pod is allowed to connect to for that mode.
+type EgressRouterModeConfig struct {
+	// Mode depicts the mode that is used for the egress router. The default mode is "Redirect" and is the only supported mode currently.
+	// +kubebuilder:validation:Required
+	// +kubebuilder:validation:Enum="Redirect"
+	// +kubebuilder:default:="Redirect"
+	Mode EgressRouterMode `json:"mode"`
+
+	// List of destinations that the pod is allowed to connect to in redirect mode.
+	// If unspecified, the router will not be configured to connect to specific remote destination.
+	// Therefore if the mode is "Redirect" then this field must be set.
+	// +kubebuilder:validation:Optional
+	AllowedDestinationsRedirect []RedirectConfig `json:"allowedDestinationsRedirect"`
+}
+
+// EgressRouterMode defines the different types of modes that are supported for the egress router interface.
+// The default mode is "Redirect" and is the only supported mode currently.
+type EgressRouterMode string
+
+const (
+	// EgressRouterModeRedirect creates an egress router that sets up iptables rules to redirect traffic
+	// from its own IP address to one or more remote destination IP addresses.
+	EgressRouterModeRedirect EgressRouterMode = "Redirect"
+)
+
+// RedirectConfig consists of parameters used to define the remote destination in the redirect mode.
+type RedirectConfig struct {
+	// IP specifies the remote destination's IP address. Can be IPv4 or IPv6.
+	// +kubebuilder:validation:Required
+	IP CIDR `json:"ip"`
+
+	// PortProtocolConfig allows the user to define the port, protocol and targetPort fields for the remote destination.
+	// +kubebuilder:validation:Optional
+	PortProtocolConfig PortProtocolRedirectConfig `json:"portProtocolConfig,omitempty"`
+}
+
+// PortProtocolRedirectConfig defines the port, protocol and targetPort fields for the remote destination in redirect mode.
+type PortProtocolRedirectConfig struct {
+	// Port allows specifying the port number on the router.
```

Comment:
Note to reviewers: this has been split out of `RedirectConfig` purposefully to ensure more control on the fields. For example these fields have to be optional from up top but once a `port` is specified then we need the `protocol` also to be specified and vice versa and so they both have to  be required fields as well.